api/client_test.go

```diff
@@ -47,46 +75,347 @@ func Test_DefaultClientConfig_WithEnvVars(t *testing.T) {
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("invalid host", func(t *testing.T) {
+		clientConfig := &ClientConfig{URL: "http://"}
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("missing cert", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.ClientCert = "../testutils/certs/localhost_cert.pem"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("missing key", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.ClientKey = "../testutils/certs/localhost_key.pem"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("invalid cert", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.ClientCert = "foo"
+		clientConfig.TLSConfig.ClientKey = "../testutils/certs/localhost_key.pem"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("invalid key", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.ClientCert = "../testutils/certs/localhost_cert.pem"
+		clientConfig.TLSConfig.ClientKey = "bar"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("invalid CA cert", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.CACert = "foo"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+
+	t.Run("invalid CA path", func(t *testing.T) {
+		clientConfig := BaseClientConfig()
+		clientConfig.TLSConfig.CAPath = "bar"
+
+		c, err := NewClient(clientConfig, nil)
+
+		assert.Nil(t, c)
+		assert.Error(t, err)
+	})
+}
+
+func Test_Client_Status(t *testing.T) {
+	c := &Client{}
+	assert.NotNil(t, c.Status())
+}
+
+func Test_QueryParam_Encode(t *testing.T) {
+	type fields struct {
+		IncludeEvents bool
+		Status        string
+		Run           string
+	}
+
+	tests := []struct {
+		name   string
+		fields fields
+		want   string
+	}{
+		{
+			name:   "empty",
+			fields: fields{},
+			want:   "",
+		},
+		{
+			name:   "single field",
+			fields: fields{Status: "foo"},
+			want:   "status=foo",
+		},
+		{
+			name:   "multiple fields",
+			fields: fields{Status: "foo", Run: "bar"},
+			want:   "run=bar&status=foo",
+		},
+		{
+			name:   "multiple fields with include events",
+			fields: fields{Status: "foo", Run: "bar", IncludeEvents: true},
+			want:   "include=events&run=bar&status=foo",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			q := &QueryParam{
+				IncludeEvents: tt.fields.IncludeEvents,
+				Status:        tt.fields.Status,
+				Run:           tt.fields.Run,
+			}
+			assert.Equalf(t, tt.want, q.Encode(), "Encode()")
+		})
+	}
+}
+
+func Test_StatusClient_Overall_BadResponse(t *testing.T) {
```

Comment:
Looks like there's a series of `Test_StatusClient_Overall_*` test cases with different test-servers. Suggestion - it seems like it would be possible to reorganize these into a single test func with a test-table that has the server as a field. I think it would reduce code and unify these test cases in a single function.

e.g.
```
cases := []struct {
	name   string
	server *httptest.Server
} {
	{
		"error with errorr message",
		httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			e := &ErrorResponse{Error: &ErrorObject{Message: "foo"}}
			bytes, err := json.Marshal(e)
			assert.NotEmpty(t, bytes)
			assert.NoError(t, err)

			w.WriteHeader(http.StatusInternalServerError)
			_, err = w.Write(bytes)
			assert.NoError(t, err)
		})),
	},
...
```

Reply:
I will try that. Initially I didn't do it because I thought the individual tests are a little too big, but putting the server in the table is a good idea.

Reply:
Addressed in https://github.com/hashicorp/consul-terraform-sync/pull/665/commits/f5d431cceb101755a27a5dffa9b828e918129a3f

Reply:
Awesome, looks good. Thanks for trying it out.